pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,147 @@
+func NewWatcher(sockDir string) Watcher {
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCbkFn),
+		stopCh:   make(chan interface{}),
+	}
+
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCbkFn) {
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					client, conn, err := dial(event.Name)
+					if err != nil {
+						glog.Warningf("Dial failed at socket %s, err: %v", event.Name, err)
+						if conn != nil {
+							conn.Close()
+						}
+						continue
+					}
+					resp, err := client.Info(context.Background(), &watcherapi.Empty{})
+					if err != nil {
+						glog.Errorf("Info RPC returned err: %v. Skip handler for %s", err, event.Name)
+						conn.Close()
+						continue
+					}
+					if handlerCbkFn, ok := w.handlers[resp.Type]; ok {
+						err = handlerCbkFn(resp, event.Name)
+						var status watcherapi.RegistrationStatus
+						if err != nil {
+							status.PluginRegistered = false
+							status.Error = fmt.Sprintf("Plugin registration failed with err: %v", err)
+						} else {
+							status.PluginRegistered = true
+						}
+						_, err = client.UpdateRegistrationStatus(context.Background(), &status)
+						if err != nil {
+							glog.Errorf("Failed to notify plugin about registration status")
+						}
+					} else {
+						glog.Errorf("No handler found registered for plugin type: %s, socket: %s", resp.Type, event.Name)
+					}
+					conn.Close()
+					continue
+				}
+			case <-w.stopCh:
+				watcher.Close()
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+// Dial establishes the gRPC communication with the picked up plugin socket. https://godoc.org/google.golang.org/grpc#Dial
+func dial(unixSocketPath string) (watcherapi.IdentityClient, *grpc.ClientConn, error) {
+	c, err := grpc.Dial(unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(10*time.Second),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to dial plugin identity service, err: %v", err)
+	}
+
+	return watcherapi.NewIdentityClient(c), c, nil
+}
+
+// GetResourceNameFrom returns base resource name from the plugin socket absolute path
+func GetResourceNameFrom(sockPath string) (string, error) {
```

Comment:
Since a resourceName requires a single `/` (e.g: `nvidia.com/gpu` is valid but `nvidia.com/gpu/memory` isnt) we could require your device plugin to create a domain name directory (`nvidia.com`) but not impose any convention on the socket name (other than `[NAME].sock`?).